Added C++ API test

izdeby opened the pull request:
This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add a test to check that C++ API behavior stays the same after all the changes. 
While working on it a bug related to `requires_grad` was found and logged in the master task.


--------------
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30983 Updated codegen to remove TensorOptions from generated code
* #30982 Prepare templates
* #30981 Updated DispatchKeyExtractor to expect TensorOptions
* **#30980 Added C++ API test**
* #30979 Add tracing support for optional Device and Layout

Differential Revision: [D18912681](https://our.internmc.facebook.com/intern/diff/D18912681)

izdeby committed:
Added C++ API test

[ghstack-poisoned]

gchanan commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -2,6 +2,7 @@
 
 #include <ATen/ATen.h>
 #include <ATen/core/Reduction.h>
+#include <torch/cuda.h>
```
what do you need from this?

izdeby replied:
for `torch::cuda::is_available()`

gchanan replied:
can you follow what the other tests in this directory do?  Namely, use `at::cuda::is_available()` (and don't include this file).

izdeby replied:
I cant use `at::cuda::is_available()` here as i will get an error: `error: ‘is_available’ is not a member of ‘at::cuda’`. Other files that use `at::cuda::is_available()` are including `<ATen/cuda/CUDAContext.h>` that enables this. I picked `<torch/cuda.h>` as its much easier to add than `<ATen/cuda/CUDAContext.h>` which will require additional work.
I can leave it as it is or move the test to another file, possibly CUDA test where its already enabled. What do you prefer?

gchanan replied:
why does that involve additional work?  You should do what the other tests do here or if your way is better, change the other tests to do things the new way.  Not have two ways of doing the same thing.

gchanan commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
+  ASSERT_FALSE(tensor0.requires_grad());
+  ASSERT_FALSE(tensor0.is_pinned());
+
+  // Test setting grad to FALSE
```
nit: Test setting requires_grad to false.

gchanan commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
+  ASSERT_FALSE(tensor0.requires_grad());
+  ASSERT_FALSE(tensor0.is_pinned());
+
+  // Test setting grad to TRUE
```
requires_grad.

gchanan commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
+    // Issue #30405
+    ASSERT_ANY_THROW(tensor1.is_pinned());
+  }
+}
```
newline.

gchanan commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
+  ASSERT_EQ(tensor0.layout(), at::kStrided);
+  ASSERT_EQ(tensor0.device(), at::kCPU);
+  // This is a bug. Requires_grad was set to TRUE but this is being ignored.
+  // Issue #30405
```
include the entire github link.

gchanan commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
+    ASSERT_FALSE(tensor1.requires_grad());
+
+    // This will cause an exception. 
+    // Issue #30405
```
include the entire github link.

gchanan commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -345,3 +346,69 @@ TEST(BasicTest, BasicTestCUDA) {
+    ASSERT_EQ(tensor1.dtype(), at::kHalf);
+    ASSERT_EQ(tensor1.layout(), at::kSparse);
+    ASSERT_TRUE(tensor1.device().is_cuda());
+    ASSERT_FALSE(tensor1.requires_grad());
```
relink the broken issue here.

izdeby committed:
Update on "Added C++ API test"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add a test to check that C++ API behavior stays the same after all the changes. 
While working on it a bug related to `requires_grad` was found and logged in the master task.


--------------



[ghstack-poisoned]

izdeby committed:
Update on "Added C++ API test"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add a test to check that C++ API behavior stays the same after all the changes. 
While working on it a bug related to `requires_grad` was found and logged in the master task.


--------------

Differential Revision: [D18912681](https://our.internmc.facebook.com/intern/diff/D18912681)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/test/basic.cpp:
```diff
@@ -345,3 +346,72 @@ TEST(BasicTest, BasicTestCUDA) {
+    ASSERT_TRUE(tensor1.device().is_cuda());
+
+    // This is a bug
+    // Issue https://github.com/pytorch/pytorch/issues/30405
```
Oof what a nasty bug
